### PR TITLE
wip feat: Ability to back universal blockstore by CQL servers

### PR DESCRIPTION
**Pull request by magik6k**
## Proposed Changes
Experiment to see if this can work

## Additional Info
* Set `LOTUS_PATH` to a new dir, or move .lotus somewhere else
* Set `LOTUS_CASSANDRA_UNIVERSAL_STORE=[connect string (usually just ip)]`
* Import snapshot / start daemon as usual

Didn't test at all yet

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: node/modules/blockstore.go [hunk 1]
```diff
@@ -23,6 +24,25 @@ import (
...
+		if err != nil {
+			return nil, err
+		}
+		copt := bstore.DefaultCacheOpts()
+		copt.HasARCCacheSize = 1 << 20
+
+		rbs := bstore.NewBlockstoreNoPrefix(cds)
+
+		// only metadata cache
+		cbs, err := bstore.CachedBlockstore(mctx, rbs, copt)
```
Comment (magik6k) [inline node/modules/blockstore.go:38]: Probably will need a real block caching layer

File: blockstore/cassbs/cassandra_ds.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+func (c *cassandraBatch) Delete(ctx context.Context, key datastore.Key) error {
+	statement := "DELETE FROM lotus.chain WHERE key = ?"
+	c.batch.Query(statement, toCasKey(key))
+	return nil
+}
+
+func (c *cassandraBatch) Commit(ctx context.Context) error {
+	var err error
+	for i := 0; i < 30; i++ {
```
Comment (snissn) [inline blockstore/cassbs/cassandra_ds.go:210]: what's the meaning of looping 30 times here?

Comment (magik6k) [reply blockstore/cassbs/cassandra_ds.go:210]: This just retries very hard

File: blockstore/cached.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+type CachedBlockstore struct {
+	write Blockstore
+	cache *lru.ARCCache[cid.Cid, block.Block]
+
+	writeLk       sync.Mutex
+	pendingWrites map[cid.Cid]block.Block
+}
+
+var CacheBstoreSize = (4 << 30) / 16000 // 4GB with average block size of 16KB
```
Comment (magik6k) [inline blockstore/cached.go:19]: Could use tuning, or at least a config

File: blockstore/cached.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	block "github.com/ipfs/go-libipfs/blocks"
+	"sync"
+)
+
+type CachedBlockstore struct {
+	write Blockstore
+	cache *lru.ARCCache[cid.Cid, block.Block]
+
+	writeLk       sync.Mutex
+	pendingWrites map[cid.Cid]block.Block
```
Comment (magik6k) [inline blockstore/cached.go:16]: This needs to be bounded

**Comment by magik6k** [comment]
This will be merged through https://github.com/filecoin-project/lotus/pull/10821

**Pull request closed without merging**